File: windows-app/Plantoir/Services/ClaudeCodeLauncher.cs
using System;
using System.Diagnostics;
using System.IO;
using System.Text;
using System.Text.Json;

namespace Plantoir.Services;

/// <summary>
/// Starting a Claude Code session already connected to this working folder's
/// Plantoir tools, locked to one course.
///
/// The teacher never types a command. Everything the connection needs is
/// written for them and thrown away with the session:
///
/// * **No global configuration is touched.** The connection is passed with
///   `--mcp-config`, and `--strict-mcp-config` means only that server is
///   loaded — so a teacher's own MCP servers are neither used nor disturbed,
///   and nothing is left behind when the session ends.
/// * **Nothing lands in the teacher's folder.** The config sits in the app's
///   own data directory, not in the vault Obsidian is watching.
/// * **The session is locked to the course it was started from.** Passed to
///   the server rather than asked for in a prompt, so it holds however the
///   conversation wanders.
///
/// The menu item only appears when this returns true from
/// <see cref="IsAvailable"/> — a teacher without Claude Code should not be
/// offered a door that opens onto an error.
/// </summary>
public static class ClaudeCodeLauncher
{
    /// <summary>Both halves have to be present: the assistant, and the tools for it to use.</summary>
    public static bool IsAvailable => FindClaude() is not null && FindServer() is not null;

    /// <summary>
    /// Where Claude Code is, or null. Checked on PATH first, then the places
    /// its own installers put it.
    /// </summary>
    public static string? FindClaude()
    {
        string? onPath = OnPath("claude.exe") ?? OnPath("claude.cmd") ?? OnPath("claude.bat");
        if (onPath is not null) return onPath;

        string home = Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        foreach (string candidate in new[]
                 {
                     Path.Combine(home, ".local", "bin", "claude.exe"),
                     Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
                         "Programs", "claude", "claude.exe"),
                     Path.Combine(home, "AppData", "Roaming", "npm", "claude.cmd"),
                 })
            if (File.Exists(candidate)) return candidate;
        return null;
    }

    /// <summary>The MCP server, which ships beside the app.</summary>
    public static string? FindServer()
    {
        string? beside = Path.GetDirectoryName(Environment.ProcessPath);
        if (beside is null) return null;
        foreach (string candidate in new[]
                 {
                     Path.Combine(beside, "plantoir-mcp.exe"),
                     // Running from a dev build, where each project has its own output.
                     Path.GetFullPath(Path.Combine(beside, "..", "..", "..", "..",
                         "Plantoir.Mcp", "bin", "Debug", "net9.0", "win-x64", "plantoir-mcp.exe")),
                 })
            if (File.Exists(candidate)) return candidate;
        return null;
    }

    private static string? OnPath(string name)
    {
        string paths = Environment.GetEnvironmentVariable("PATH") ?? "";
        foreach (string directory in paths.Split(Path.PathSeparator, StringSplitOptions.RemoveEmptyEntries))
        {
            try
            {
                string candidate = Path.Combine(directory.Trim(), name);
                if (File.Exists(candidate)) return candidate;
            }
            catch { }
        }
        return null;
    }

    /// <summary>
    /// Open a session for one course. Returns false when it could not be
    /// started, so the caller can say so rather than leaving a teacher looking
    /// at nothing.
    /// </summary>
    public static bool Open(string workspacePath, string courseCode, string courseName)
    {
        string? claude = FindClaude();
        string? server = FindServer();
        if (claude is null || server is null) return false;

        string configPath;
        try { configPath = WriteConfig(workspacePath, courseCode, server); }
        catch { return false; }

        string prompt = Greeting(courseCode, courseName);
        string command =
            $"\"{claude}\" --mcp-config \"{configPath}\" --strict-mcp-config \"{prompt}\"";

        // Windows Terminal when it is there, because a teacher will be reading
        // this for a while; the classic console otherwise.
        var info = new ProcessStartInfo { UseShellExecute = true, WorkingDirectory = workspacePath };
        if (OnPath("wt.exe") is not null)
        {
            info.FileName = "wt.exe";
            info.Arguments = $"-d \"{workspacePath}\" cmd /k {command}";
        }
        else
        {
            info.FileName = "cmd.exe";
            info.Arguments = $"/k {command}";
        }

        try { return Process.Start(info) is not null; }
        catch { return false; }
    }

    /// <summary>
    /// The opening message. It names the course and points at the plan tools,
    /// because the safety of every write here depends on a plan being shown to
    /// the teacher first — and an assistant that starts by reading is far more
    /// useful than one that starts by asking what to do.
    /// </summary>
    private static string Greeting(string courseCode, string courseName)
    {
        var text = new StringBuilder();
        text.Append($"I'm a teacher working on {courseCode}");
        if (!string.IsNullOrWhiteSpace(courseName) && courseName != courseCode)
            text.Append($" ({courseName})");
        text.Append(" in Plantoir. Use the plantoir tools for anything to do with this course. ");
        text.Append("Start by listing its sections so we both know what's there. ");
        text.Append("Before changing anything, use the matching plan tool first and show me what it says, ");
        text.Append("in plain words, and wait for me to agree.");
        return text.ToString().Replace("\"", "'");
    }

    /// <summary>
    /// The connection, written per course into the app's own data directory.
    /// One file per course so two sessions on different courses do not
    /// overwrite each other's configuration mid-launch.
    /// </summary>
    private static string WriteConfig(string workspacePath, string courseCode, string server)
    {
        string directory = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "Plantoir", "assist");
        Directory.CreateDirectory(directory);

        var config = new
        {
            mcpServers = new Dictionary<string, object>
            {
                ["plantoir"] = new
                {
                    command = server,
                    args = new[] { "--folder", workspacePath, "--course", courseCode },
                },
            },
        };

        string path = Path.Combine(directory, $"mcp-{courseCode}.json");
        File.WriteAllText(path, JsonSerializer.Serialize(config,
            new JsonSerializerOptions { WriteIndented = true }));
        return path;
    }
}

File: mac-app/QuartzTeachers/Models/Assist/ClaudeCodeLauncher.swift
import AppKit
import Foundation

/// Starting a Claude Code session already connected to this working folder's
/// Plantoir tools, locked to one course.
///
/// The teacher never types a command. Everything the connection needs is
/// written for them and thrown away with the session:
///
/// * **No global configuration is touched.** The connection is passed with
///   `--mcp-config`, and `--strict-mcp-config` means only that server is
///   loaded — so a teacher's own MCP servers are neither used nor disturbed,
///   and nothing is left behind when the session ends.
/// * **Nothing lands in the teacher's folder.** The config sits in the app's
///   own data directory, not in the vault Obsidian is watching.
/// * **The session is locked to the course it was started from.** Passed to
///   the server rather than asked for in a prompt, so it holds however the
///   conversation wanders.
///
/// The menu item only appears when this returns true from
/// `isAvailable` — a teacher without Claude Code should not be offered a
/// door that opens onto an error.
nonisolated enum ClaudeCodeLauncher {

    // MARK: - Computed properties

    /// Both halves have to be present: the assistant, and the tools for it to use.
    static var isAvailable: Bool {
        return findClaude() != nil && findServer() != nil
    }

    // MARK: - Functions

    /// Where Claude Code is, or nil. Checked on PATH first, then the places
    /// its own installers put it.
    static func findClaude() -> String? {
        if let onPath = onPath("claude") {
            return onPath
        }

        let homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
        var candidates: [String] = [
            homeDirectory.appendingPathComponent(".local/bin/claude").path,
            "/opt/homebrew/bin/claude",
            "/usr/local/bin/claude",
            homeDirectory.appendingPathComponent(".npm-global/bin/claude").path,
            homeDirectory.appendingPathComponent(".bun/bin/claude").path,
        ]

        let nvmNodeDirectory: URL = homeDirectory.appendingPathComponent(".nvm/versions/node")
        if let nodeVersions = try? FileManager.default.contentsOfDirectory(atPath: nvmNodeDirectory.path) {
            for version in nodeVersions {
                let nvmClaudePath: String = nvmNodeDirectory.appendingPathComponent("\(version)/bin/claude").path
                candidates.append(nvmClaudePath)
            }
        }

        for candidate in candidates {
            if FileManager.default.isExecutableFile(atPath: candidate) {
                return candidate
            }
        }
        return nil
    }

    /// The MCP server, which is the Plantoir binary itself answering `--mcp-stdio`.
    static func findServer() -> String? {
        if let executablePath = Bundle.main.executablePath,
           FileManager.default.isExecutableFile(atPath: executablePath) {
            return executablePath
        }
        if let firstArgument = ProcessInfo.processInfo.arguments.first,
           FileManager.default.isExecutableFile(atPath: firstArgument) {
            return firstArgument
        }
        return nil
    }

    /// Check if a binary exists and is executable in any folder listed in PATH.
    static func onPath(_ name: String) -> String? {
        guard let pathVariable = ProcessInfo.processInfo.environment["PATH"] else {
            return nil
        }
        let directories: [String] = pathVariable.components(separatedBy: ":")
        for directory in directories {
            let candidate: String = (directory.trimmingCharacters(in: .whitespacesAndNewlines) as NSString)
                .appendingPathComponent(name)
            if FileManager.default.isExecutableFile(atPath: candidate) {
                return candidate
            }
        }
        return nil
    }

    /// Open a session for one course. Returns false when it could not be
    /// started, so the caller can say so rather than leaving a teacher looking
    /// at nothing.
    @discardableResult
    static func open(workspacePath: String, courseCode: String, courseName: String) -> Bool {
        guard let claude: String = findClaude(),
              let server: String = findServer() else {
            return false
        }

        let configPath: String
        do {
            configPath = try writeConfig(workspacePath: workspacePath, courseCode: courseCode, serverPath: server)
        } catch {
            return false
        }

        let prompt: String = greeting(courseCode: courseCode, courseName: courseName)

        let scriptPath: String
        do {
            scriptPath = try writeLauncherScript(
                workspacePath: workspacePath,
                courseCode: courseCode,
                claudePath: claude,
                configPath: configPath,
                prompt: prompt
            )
        } catch {
            return false
        }

        ActivityTrail.note(.assistantOpened, "started Claude Code for \(courseCode)")

        return launchTerminal(scriptPath: scriptPath)
    }

    /// The opening message. It names the course and points at the plan tools,
    /// because the safety of every write here depends on a plan being shown to
    /// the teacher first — and an assistant that starts by reading is far more
    /// useful than one that starts by asking what to do.
    static func greeting(courseCode: String, courseName: String) -> String {
        var text: String = "I'm a teacher working on \(courseCode)"
        let trimmedName: String = courseName.trimmingCharacters(in: .whitespacesAndNewlines)
        if !trimmedName.isEmpty && trimmedName != courseCode {
            text.append(" (\(trimmedName))")
        }
        text.append(" in Plantoir. Use the plantoir tools for anything to do with this course. ")
        text.append("Start by listing its sections so we both know what's there. ")
        text.append("Before changing anything, use the matching plan tool first and show me what it says, ")
        text.append("in plain words, and wait for me to agree.")
        return text.replacingOccurrences(of: "\"", with: "'")
    }

    /// The connection, written per course into the app's own data directory.
    /// One file per course so two sessions on different courses do not
    /// overwrite each other's configuration mid-launch.
    static func writeConfig(workspacePath: String, courseCode: String, serverPath: String) throws -> String {
        let appSupportDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Plantoir/assist")
        try FileManager.default.createDirectory(at: appSupportDirectory, withIntermediateDirectories: true)

        let configuration: [String: Any] = [
            "mcpServers": [
                "plantoir": [
                    "command": serverPath,
                    "args": [
                        "--mcp-stdio",
                        workspacePath
                    ]
                ]
            ]
        ]

        let data: Data = try JSONSerialization.data(
            withJSONObject: configuration,
            options: [.prettyPrinted, .sortedKeys]
        )
        let targetFile: URL = appSupportDirectory.appendingPathComponent("mcp-\(courseCode).json")
        try data.write(to: targetFile)
        return targetFile.path
    }

    /// Writes an executable `.command` launcher script in the app's data directory.
    static func writeLauncherScript(
        workspacePath: String,
        courseCode: String,
        claudePath: String,
        configPath: String,
        prompt: String
    ) throws -> String {
        let appSupportDirectory: URL = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent("Library/Application Support/Plantoir/assist")
        try FileManager.default.createDirectory(at: appSupportDirectory, withIntermediateDirectories: true)

        let scriptContent: String = """
        #!/bin/bash
        cd \(escapeForShell(workspacePath)) || exit 1
        \(escapeForShell(claudePath)) --mcp-config \(escapeForShell(configPath)) --strict-mcp-config \(escapeForShell(prompt))
        """

        let scriptFile: URL = appSupportDirectory.appendingPathComponent("launch-\(courseCode).command")
        try scriptContent.write(to: scriptFile, atomically: true, encoding: .utf8)

        try FileManager.default.setAttributes([.posixPermissions: 0o755], ofItemAtPath: scriptFile.path)
        return scriptFile.path
    }

    /// Single-quote a string safely for POSIX shell execution.
    static func escapeForShell(_ text: String) -> String {
        return "'" + text.replacingOccurrences(of: "'", with: "'\\''") + "'"
    }

    /// Launches a terminal emulator running the generated .command script.
    ///
    /// Uses LaunchServices via `NSWorkspace` instead of AppleScript/AppleEvents
    /// so macOS does not require Automation / AppleEvents permissions.
    private static func launchTerminal(scriptPath: String) -> Bool {
        let itermBundleID: String = "com.googlecode.iterm2"
        let isITermRunning: Bool = !NSRunningApplication.runningApplications(
            withBundleIdentifier: itermBundleID
        ).isEmpty

        if isITermRunning {
            if NSWorkspace.shared.openFile(scriptPath, withApplication: "iTerm") {
                return true
            }
        }

        if NSWorkspace.shared.openFile(scriptPath, withApplication: "Terminal") {
            return true
        }

        let scriptURL: URL = URL(fileURLWithPath: scriptPath)
        return NSWorkspace.shared.open(scriptURL)
    }
}
